File: Sources/CreatorCoreForge/BookImporter.swift
import Foundation

/// Result of importing a text file into chapters with metadata.
public struct ImportedBook {
    public let title: String
    public let author: String?
    public let chapters: [Chapter]

    public init(title: String, author: String?, chapters: [Chapter]) {
        self.title = title
        self.author = author
        self.chapters = chapters
    }
}

/// Parses raw text into `Chapter` objects by blank lines.
public struct ChapterParser {
    public static func parse(_ text: String) -> [Chapter] {
        let normalized = text.replacingOccurrences(of: "\r\n", with: "\n")
        let blocks = normalized.components(separatedBy: "\n\n")
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        var chapters: [Chapter] = []
        var idx = 1
        for b in blocks {
            let lines = b.split(separator: "\n")
            guard let first = lines.first else { continue }
            var title = String(first).trimmingCharacters(in: .whitespaces)
            var start = 1
            if !title.lowercased().hasPrefix("chapter") {
                title = "Chapter \(idx)"
                start = 0
            }
            let body = lines.dropFirst(start).joined(separator: " ")
            if body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { continue }
            chapters.append(Chapter(title: title, text: body))
            idx += 1
        }
        return chapters
    }
}

/// Imports TXT, PDF, and EPUB files from disk.
public final class BookImporter {
    public init() {}

    public func `import`(fileURL: URL) throws -> ImportedBook {
        let ext = fileURL.pathExtension.lowercased()
        let raw = try String(contentsOf: fileURL)
        let meta = Self.extractMetadata(from: raw, fileName: fileURL.lastPathComponent)
        let body = meta.body
        let chapters: [Chapter]
        switch ext {
        case "txt":
            chapters = ChapterParser.parse(body)
        case "pdf", "epub":
            chapters = [Chapter(title: "Chapter 1", text: body.replacingOccurrences(of: "\n", with: " "))]
        default:
            throw NSError(domain: "BookImporter", code: 1)
        }
        return ImportedBook(title: meta.title, author: meta.author, chapters: chapters)
    }

    private static func extractMetadata(from text: String, fileName: String) -> (title: String, author: String?, body: String) {
        let lines = text.replacingOccurrences(of: "\r\n", with: "\n").split(separator: "\n")
        var title = fileName.replacingOccurrences(of: "." + fileName.split(separator: ".").last!, with: "")
        var author: String?
        var index = 0
        if index < lines.count, lines[index].lowercased().hasPrefix("title:") {
            title = lines[index].dropFirst(6).trimmingCharacters(in: .whitespaces)
            index += 1
        } else if index < lines.count, lines[index].hasPrefix("#") {
            title = lines[index].trimmingCharacters(in: CharacterSet(charactersIn: "# "))
            index += 1
        }
        if index < lines.count, lines[index].lowercased().hasPrefix("author:") {
            author = lines[index].dropFirst(7).trimmingCharacters(in: .whitespaces)
            index += 1
        }
        let body = lines.dropFirst(index).joined(separator: "\n")
        return (title, author, body)
    }
}

File: VoiceLab/src/bookImporter.ts
export interface Chapter {
  title: string;
  text: string;
}

export interface ImportedBook {
  title: string;
  author?: string;
  chapters: Chapter[];
}

function normalize(text: string): string {
  return text.replace(/\s+/g, ' ').trim();
}

export class ChapterParser {
  static parse(text: string): Chapter[] {
    const blocks = text.replace(/\r\n/g, '\n').split(/\n\s*\n/).filter(b => b.trim());
    const chapters: Chapter[] = [];
    let idx = 1;
    for (const b of blocks) {
      const lines = b.trim().split(/\n+/);
      if (!lines.length) continue;
      let title = lines[0].trim();
      let start = 1;
      if (!/^chapter\b/i.test(title)) {
        title = `Chapter ${idx}`;
        start = 0;
      }
      const body = normalize(lines.slice(start).join(' '));
      if (!body) continue;
      chapters.push({ title, text: body });
      idx++;
    }
    return chapters;
  }
}

function extractMetadata(text: string, fileName: string): { title: string; author?: string; body: string } {
  const lines = text.replace(/\r\n/g, '\n').split('\n');
  let title = fileName.replace(/\.[^.]+$/, '');
  let author: string | undefined;
  let start = 0;
  if (lines[start] && /^title:/i.test(lines[start])) {
    title = lines[start].replace(/^title:/i, '').trim();
    start++;
  } else if (lines[start] && lines[start].startsWith('#')) {
    title = lines[start].replace(/^#+/, '').trim();
    start++;
  }
  if (lines[start] && /^author:/i.test(lines[start])) {
    author = lines[start].replace(/^author:/i, '').trim();
    start++;
  }
  const body = lines.slice(start).join('\n');
  return { title, author, body };
}

export class BookImporter {
  async import(file: File): Promise<ImportedBook> {
    const ext = file.name.split('.').pop()?.toLowerCase();
    const raw = await file.text();
    const { title, author, body } = extractMetadata(raw, file.name);
    let chapters: Chapter[];
    if (ext === 'txt') {
      chapters = ChapterParser.parse(body);
    } else if (ext === 'pdf' || ext === 'epub') {
      chapters = [{ title: 'Chapter 1', text: normalize(body) }];
    } else {
      throw new Error('Unsupported format');
    }
    return { title, author, chapters };
  }
}

// backwards compatibility
export async function importBook(file: File): Promise<Chapter[]> {
  const importer = new BookImporter();
  const result = await importer.import(file);
  return result.chapters;
}
